### PR TITLE
Acceptance tests no root floc 4316

### PR DESCRIPTION
**Pull request by justnoise**
Well this got out of hand quickly...  There was a smaller PR but our docs are generated from our provisioning code so it's necessary to update those concurrently.  Since this is a large change, I'm very happy to walk through the changes with the reviewer.

This PR exists to support a new tool that allows our users to use our provisioning code to provision a cluster easily.  To ensure a user doesn't need to allow root to ssh into their machine, the provisioning code has been changed to run everything through sudo.  (yes, this assumes that the user of the provisioning tool will have passwordless sudo enabled).

A few new Effect classes have been added for the sudo commands, the nastiest of which is sudo_script.  This packages multiple bash commands into a shell quoted string and runs them using `bash -c 'commands'`.

### CODE REVIEW

**Comment by wallrj** [comment]
@justnoise there are some conflicts here so I'll wait for you to merge the branch forward.

File: admin/client.py [hunk 1]
```diff
@@ -279,6 +296,8 @@ def get_steps_pkg(distribution, package_source=PackageSource()):
     package_manager = DOCKER_IMAGES[distribution].package_manager
     steps = [
         ensure_minimal_setup(package_manager),
+        allow_notty_sudo(distribution),
```
Comment (justnoise) [inline admin/client.py]: Sadly we still need a tty for running a script on docker (but not for ssh)

File: admin/client.py [hunk 2]
```diff
@@ -279,6 +296,8 @@ def get_steps_pkg(distribution, package_source=PackageSource()):
...
+        install_distro_package("curl", distribution),
```
Comment (justnoise) [inline admin/client.py]: It's possible this should go in the ensure_minimal_setup but I'm unsure if we should pollute that function with installing curl.

File: flocker/provision/_install.py [hunk 3]
```diff
@@ -743,7 +754,12 @@ def _remove_private_key(content):
     if trim_end <= trim_start:
         # strangely short key, keep all content
         return content
-    return content[:trim_start] + '...REMOVED...' + content[trim_end:]
+    end_of_key = end+len(suffix)
+    trimmed_key = (content[:trim_start] +
+                   '...REMOVED...' +
+                   content[trim_end:end_of_key])
+    # keep looking for more keys
+    return trimmed_key + _remove_private_keys(content[end_of_key:])
```
Comment (justnoise) [inline flocker/provision/_install.py]: Probably uncecessary but updated this to remove multiple keys from the output (used to stop after a single key)

**Comment by justnoise** [comment]
OK this is ready for review... hopefully for good this time.

File: flocker/provision/_install.py [hunk 4]
```diff
@@ -140,6 +140,15 @@ def apt_get_install(args, sudo=False):
     )
 
 
+def install_distro_package(package_name, distribution, sudo=False):
+    if type(package_name) != list:
+        package_name = [package_name]
+    if is_centos(distribution):
```
Comment (myechuri) [inline flocker/provision/_install.py]: Address RHEL.

File: flocker/provision/_install.py [hunk 5]
```diff
@@ -915,21 +940,27 @@ def task_enable_flocker_control(distribution, action="start"):
...
                 content=dedent('''\
                     start on runlevel [2345]
                     stop on runlevel [016]
                     '''),
             ),
-            run("echo 'flocker-control-api\t4523/tcp\t\t\t# Flocker Control API port' >> /etc/services"),  # noqa
-            run("echo 'flocker-control-agent\t4524/tcp\t\t\t# Flocker Control Agent port' >> /etc/services"),  # noqa
-            run_from_args(['service', 'flocker-control', action.lower()]),
+            sudo_script(dedent("""\
+            if grep -q flocker-control-api /etc/services; then
```
Comment (myechuri) [inline flocker/provision/_install.py]: Bug: need to make sure it's NOT in /etc/services

File: flocker/provision/_install.py [hunk 6]
```diff
@@ -915,21 +940,27 @@ def task_enable_flocker_control(distribution, action="start"):
...
+                echo 'flocker-control-api\t4523/tcp\t\t\t# Flocker Control API port' >> /etc/services
+            fi""")),  # noqa
+            sudo_script(dedent("""\
+            if grep -q flocker-control-agent /etc/services; then
```
Comment (myechuri) [inline flocker/provision/_install.py]: Same Bug.

**Comment by wallrj** [comment]
Branch has merge conflicts and the certificate distribution problem is going to go away in the next version of Flocker.

**Pull request closed without merging**